js/stdlib/ts/ETH_like.ts

```diff
@@ -266,6 +266,103 @@ const stepTime = async (): Promise<TransactionReceipt> => {
   return await transfer(faucet, acc, parseCurrency(0));
 };
 
+// ****************************************************************************
+// Event Cache
+// ****************************************************************************
+
+const getMinBlock = (logs: any[]) =>
+  logs.reduce((acc: Log, x: Log) =>
+    (x.blockNumber == acc.blockNumber)
+      ? (x.logIndex < acc.logIndex ? x : acc)
+      : (x.blockNumber.toString() < acc.blockNumber.toString() ? x : acc), logs[0]);
+
+const getMaxBlock = (logs: any[]) =>
+  logs.reduce((acc: Log, x: Log) =>
+    (x.blockNumber == acc.blockNumber)
+      ? (x.logIndex > acc.logIndex ? x : acc)
+      : (x.blockNumber.toString() > acc.blockNumber.toString() ? x : acc), logs[0]);
+
+
+class EventCache {
+
+  cache: any[] = [];
+
+  currentBlock = 0;
+
+  constructor() {
+    this.cache = [];
+  }
+
+  async query_(fromBlock: number, toBlock: number, topic: string, getLogs: (currentBlock: number) => Promise<any[]>) {
+    debug(`EventCache.query`, fromBlock, toBlock, topic);
+    if (fromBlock > toBlock) {
+      return undefined;
+    }
+    // Clear cache of stale transactions
+    // Cache's min bound will be `fromBlock`
+    this.cache = this.cache.filter((x) => x.blockNumber >= fromBlock);
+
+    // When checking for topic, only choose blocks that are lte `toBlock`
+    const filterFn = (x: any) =>
+      x.topics.includes(topic.toString())
+      && x.blockNumber <= toBlock;
+
+    // Check to see if the transaction we want is in the cache
+    const initLogs = this.cache.filter(filterFn);
+    if(initLogs.length > 0) {
+      debug(`Found transaction in Event Cache`);
+      return getMinBlock(initLogs);
+    }
+
+    debug(`Transaction not in Event Cache. Querying network...`);
+
+    // If no results, then contact network
+    this.cache = await getLogs(this.currentBlock + 1);
+
+    this.currentBlock =
+      (this.cache.length == 0)
+        ? toBlock
```

Comment:
I wonder if `getLogs` is hiding anything from us, such as the actual time of the query? At the very least, you could get the current time BEFORE the `getLogs` call and use that as the "current time" analog from ALGO.

Also, it is possible that we queried for `[200, 1000]` but the real network time is `500`, so you're setting it to `1000`, when you should be setting it to `500`.